spanner/client_test.go

```diff
@@ -591,6 +591,97 @@ func testReadOnlyTransaction(t *testing.T, executionTimes map[string]SimulatedEx
 	return executeSingerQuery(ctx, tx)
 }
 
+func TestClient_ReadWriteTransaction_StatementBased(t *testing.T) {
+	t.Parallel()
+
+	rowCount, attempts, err := testReadWriteTransactionStatementBased(t, make(map[string]SimulatedExecutionTime))
+	if err != nil {
+		t.Fatalf("transaction failed to commit: %v", err)
+	}
+	if rowCount != SelectSingerIDAlbumIDAlbumTitleFromAlbumsRowCount {
+		t.Fatalf("Row count mismatch, got %v, expected %v", rowCount, SelectSingerIDAlbumIDAlbumTitleFromAlbumsRowCount)
+	}
+	if g, w := attempts, 1; g != w {
+		t.Fatalf("number of attempts mismatch:\nGot%d\nWant:%d", g, w)
+	}
+}
+
+func TestClient_ReadWriteTransaction_StatementBased_CommitAborted(t *testing.T) {
+	t.Parallel()
+	rowCount, attempts, err := testReadWriteTransactionStatementBased(t, map[string]SimulatedExecutionTime{
+		MethodCommitTransaction: {Errors: []error{status.Error(codes.Aborted, "Transaction aborted")}},
+	})
+	if err != nil {
+		t.Fatalf("transaction failed to commit: %v", err)
+	}
+	if rowCount != SelectSingerIDAlbumIDAlbumTitleFromAlbumsRowCount {
+		t.Fatalf("Row count mismatch, got %v, expected %v", rowCount, SelectSingerIDAlbumIDAlbumTitleFromAlbumsRowCount)
+	}
+	if g, w := attempts, 2; g != w {
+		t.Fatalf("number of attempts mismatch:\nGot%d\nWant:%d", g, w)
+	}
+}
+
+func testReadWriteTransactionStatementBased(t *testing.T, executionTimes map[string]SimulatedExecutionTime) (int64, int, error) {
+	server, client, teardown := setupMockedTestServer(t)
+	defer teardown()
+	for method, exec := range executionTimes {
+		server.TestSpanner.PutExecutionTime(method, exec)
+	}
+	ctx := context.Background()
+
+	f := func(tx *ReadWriteTransactionStmtBased) (int64, error) {
+		iter := tx.Query(ctx, NewStatement(SelectSingerIDAlbumIDAlbumTitleFromAlbums))
+		defer iter.Stop()
+		rowCount := int64(0)
+		for {
+			row, err := iter.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				return 0, err
+			}
+			var singerID, albumID int64
+			var albumTitle string
+			if err := row.Columns(&singerID, &albumID, &albumTitle); err != nil {
+				return 0, err
+			}
+			rowCount++
+		}
+		return rowCount, nil
+	}
+
+	var (
+		rowCount int64
+		err      error
+		attempts int
+	)
+	// Retry with fixed time intervals.
+	for i, max := 0, 3; i < max; i++ {
```

Comment:
Is there a specific reason for retrying at most 3 times? I would normally not expect a max here at all. The 'normal' `ReadWriteTransaction` will also retry indefinitely until the transaction succeeds or fails with an error other than `Aborted`.

Reply:
Okay. For retrying, I usually think to only retry at most X times instead of hanging there forever. I have fixed it to follow the convention.